moveit_ros/moveit_servo/src/servo_node.cpp

```diff
@@ -31,133 +31,220 @@
-                         "Check the parameters YAML file used to launch this node.");
-    has_error = true;
+    servo_->setCollisionChecking(false);
+    response->message = "Servoing disabled";
   }
-
-  if (!servo_params.publish_joint_positions && !servo_params.publish_joint_velocities &&
-      !servo_params.publish_joint_accelerations)
+  else
   {
-    RCLCPP_ERROR(LOGGER, "At least one of publish_joint_positions / "
-                         "publish_joint_velocities / "
-                         "publish_joint_accelerations must be true. "
-                         "Check the parameters YAML file used to launch this node.");
-    has_error = true;
+    servo_->setCollisionChecking(true);
+    response->message = "Servoing enabled";
   }
+}
 
-  if ((servo_params.command_out_type == "std_msgs/Float64MultiArray") && servo_params.publish_joint_positions &&
-      servo_params.publish_joint_velocities)
+void ServoNode::switchCommandType(const std::shared_ptr<moveit_msgs::srv::ServoCommandType::Request> request,
+                                  const std::shared_ptr<moveit_msgs::srv::ServoCommandType::Response> response)
+{
+  const bool is_valid = (request->command_type >= 0) && (request->command_type <= 2);
+  if (is_valid)
   {
-    RCLCPP_ERROR(LOGGER, "When publishing a std_msgs/Float64MultiArray, "
-                         "you must select positions OR velocities."
-                         "Check the parameters YAML file used to launch this node.");
-    has_error = true;
+    servo_->expectedCommandType(static_cast<CommandType>(request->command_type));
+    response->expected_type = static_cast<int8_t>(servo_->expectedCommandType());
   }
-
-  if (servo_params.scene_collision_proximity_threshold < servo_params.self_collision_proximity_threshold)
+  else
   {
-    RCLCPP_ERROR(LOGGER, "Parameter 'self_collision_proximity_threshold' should probably be less "
-                         "than or equal to 'scene_collision_proximity_threshold'."
-                         "Check the parameters YAML file used to launch this node.");
-    has_error = true;
+    RCLCPP_WARN_STREAM(LOGGER, "Unknown command type " << request->command_type << " requested");
   }
+}
+
+void ServoNode::servoLoop()
+{
+  rclcpp::WallRate servo_frequency(1 / servo_params_.publish_period);
+  KinematicState next_joint_states;
 
-  if (has_error)
+  bool publish_command = false;
+
+  while (rclcpp::ok() && !stop_servo_)
   {
-    throw std::runtime_error("Servo failed to initialize : Invalid parameter values");
+    // Skip processing if servoing is disabled.
+    if (servo_paused_)
+      continue;
+
+    servo_params_ = servo_->getParams();
+    publish_command = false;
+    CommandType expectedType = servo_->expectedCommandType();
+
+    if (expectedType == CommandType::JOINT_JOG && new_joint_jog_msg_)
+    {
+      // Mark latest jointjog command as processed and also reject any twist and pose messages that had arrived simultaneously.
```

Comment:
For maintanability, I would suggest not listing everything out and saying something like

```suggestion
      // Mark latest joint jog command as processed and also reject any other command types that had arrived simultaneously.
```

and same for the other comments like it here